src/main/java/taskservice/FileSystemTaskService.java

```diff
@@ -0,0 +1,179 @@
+        this.get();
+    }
+
+    @Override
+    public Optional<Task> getById(int id) throws TaskServiceException {
+        if (!this.isValidTaskId(id)) {
+            return Optional.empty();
+        }
+        return Optional.of(this.tasks.get(id).clone());
+    }
+
+    @Override
+    public Task[] get() throws TaskServiceException {
+        if (!this.hasReadFromFileSystem) {
+            BufferedReader reader = null;
+            try {
+                Files.createDirectories(this.rootDir);
+                this.tasksDir.toFile().createNewFile();
+
+                reader = Files.newBufferedReader(this.tasksDir);
+                while (true) {
+                    final String currentLine = reader.readLine();
+                    if (currentLine == null) {
+                        break;
+                    }
+                    this.tasks.add(this.parseStringToTask(currentLine));
+                }
+            } catch (IOException ex) {
+                throw new TaskServiceException(ex.getMessage());
+            } finally {
+                this.closeReader(reader);
+            }
+            this.hasReadFromFileSystem = true;
+        }
+
+        return this.tasks.stream().map((task) -> task.clone()).toArray(Task[]::new);
+    }
+
+    @Override
+    public int getNumberOfTasks() {
+        return this.tasks.size();
+    }
+
+    @Override
+    public void create(Task taskToCreate) throws TaskServiceException {
+        this.tasks.add(taskToCreate.clone());
+        this.saveTasksToFileSystem();
+    }
+
+    @Override
+    public void update(int id, Task taskToUpdate) throws TaskServiceException {
+        if (!this.isValidTaskId(id)) {
+            throw new TaskNotFoundException();
+        }
+        this.tasks.set(id, taskToUpdate.clone());
+        this.saveTasksToFileSystem();
+    }
+
+    @Override
+    public void delete(int id) throws TaskServiceException {
+        if (!this.isValidTaskId(id)) {
+            throw new TaskNotFoundException();
+        }
+        this.tasks.remove(id);
+        this.saveTasksToFileSystem();
+    }
+
+    private Task parseStringToTask(String str) throws TaskServiceException {
+        final String[] tokens = str.split("\\|");
+        if (tokens.length < 3 || (!tokens[1].trim().equals("0") && !tokens[1].trim().equals("1"))) {
+            throw new TaskServiceException(FileSystemTaskService.EXCEPTION_INVALID_FILE_FORMAT);
+        }
+
+        Task task;
+        switch (tokens[0].trim()) {
+        case "T":
+            task = new Todo(tokens[2].trim());
+            break;
+        case "D":
+            task = new Deadline(tokens[2].trim(), LocalDate.parse(tokens[3].trim(), Deadline.DATE_INPUT_FORMAT));
+            break;
+        case "E":
+            task = new Event(tokens[2].trim(), tokens[3].trim());
+            break;
+        default:
+            throw new TaskServiceException(FileSystemTaskService.EXCEPTION_INVALID_FILE_FORMAT);
```

Comment:
Similar as above, for readability may be good to assign a variable to token[0/1/2].